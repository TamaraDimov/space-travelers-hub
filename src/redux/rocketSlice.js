import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
// import { useDispatch, useSelector } from 'react-redux';

const initialState = {
  rockets: [],
  toFetch: 'true',
};

export const fetchRockets = createAsyncThunk(
  'rockets/fetchRockets',
  async () => {
    const response = await fetch('https://api.spacexdata.com/v3/rockets');
    return response.json();
  },
);

const rocketsSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {},
  extraReducers(builder) {
    builder.addCase(fetchRockets.fulfilled, (state, action) => {
      const newState = { ...state };
      const rocketsCard = [];
      action.payload.forEach((rockCard) => {
        rocketsCard.push({
          id: rockCard.rocket_id,
          name: rockCard.rocket_name,
          description: rockCard.description,
          image: rockCard.flickr_images,
        });
      });
      newState.rockets = rocketsCard;
      newState.toFetch = false;
      return newState;
    });
  },
});

export default rocketsSlice.reducer;
