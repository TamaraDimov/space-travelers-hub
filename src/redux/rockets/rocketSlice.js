import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

const initialState = {
  rocketList: [],
  toFetch: 'true',
};

export const fetchRockets = createAsyncThunk(
  'rockets/fetchRockets',
  async () => {
    const response = await fetch('https://api.spacexdata.com/v3/rockets');
    return response.json();
  },
);

const rocketsSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {
    bookRocket: (state, action) => {
      const newState = state.rocketList.map((rocket) => {
        if (rocket.id !== action.payload) {
          return rocket;
        }
        return { ...rocket, reserved: true };
      });
      return { ...state, rocketList: newState };
    },
    leaveRocket: (state, action) => {
      const newState = state.rocketList.map((rocket) => {
        if (rocket.id !== action.payload) {
          return rocket;
        }
        return { ...rocket, reserved: false };
      });
      return { ...state, rocketList: newState };
    },
  },
  extraReducers(builder) {
    builder.addCase(fetchRockets.fulfilled, (state, action) => {
      const newState = { ...state };
      const rocketsCard = [];
      action.payload.forEach((rockCard) => {
        rocketsCard.push({
          id: rockCard.rocket_id,
          name: rockCard.rocket_name,
          description: rockCard.description,
          image: rockCard.flickr_images,
        });
      });
      newState.rocketList = rocketsCard;
      newState.toFetch = false;
      return newState;
    });
  },
});

export default rocketsSlice.reducer;
export const { bookRocket, leaveRocket } = rocketsSlice.actions;
