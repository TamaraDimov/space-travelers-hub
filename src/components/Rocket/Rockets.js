import { useDispatch } from 'react-redux';
import React from 'react';
import PropTypes from 'prop-types';
import style from '../style/style.css';
import { bookRocket, leaveRocket } from '../../redux/rockets/rocketSlice';

export default function Rockets({
  id, name, description, image, reserved,
}) {
  const dispatch = useDispatch();
  return (
    <main className={style.main}>
      <div className="card">
        <div>
          <img src={image} className="image" alt="rocket" />
        </div>
        <div className="options">
          <h3 className="name">{name}</h3>
          <h4>{reserved ? 'reserved' : ''}</h4>
          <p className="description>">{description}</p>
          <button
            type="button"
            className={reserved ? 'reserved unbook' : 'reserved book'}
            onClick={
              reserved
                ? () => {
                  dispatch(leaveRocket(id));
                }
                : () => {
                  dispatch(bookRocket(id));
                }
            }
          >
            {reserved ? 'Unbook Rocket' : 'Book Rocket'}
          </button>
        </div>
      </div>
    </main>
  );
}

Rockets.defaultProps = {
  id: '',
  name: '',
  description: '',
  image: '',
  reserved: false,
};
Rockets.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string,
  description: PropTypes.string,
  image: PropTypes.string,
  reserved: PropTypes.bool,
};
