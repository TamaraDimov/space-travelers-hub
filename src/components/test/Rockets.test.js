import React from 'react';
import { useDispatch } from 'react-redux';
import { render, create } from 'react-test-renderer';
import Rockets from '../Rocket/Rockets';
import { bookRocket, leaveRocket } from '../../redux/rockets/rocketSlice';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
}));

describe('Rockets component', () => {
  let props;
  let dispatch;
  let component;

  beforeEach(() => {
    props = {
      id: '1',
      name: 'Falcon 1',
      description:
        'The Falcon 1 was a small rocket designed for small payloads and low Earth orbits.',
      image: 'https://spacex.com/falcon-1.png',
      reserved: false,
    };
    dispatch = jest.fn();
    useDispatch.mockReturnValue(dispatch);
    component = create(<Rockets {...props} />);
  });

  it('should render correctly', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should call bookRocket action when the "Book Rocket" button is clicked', () => {
    const button = component.root.findByType('button');
    button.props.onClick();
    expect(dispatch).toHaveBeenCalledWith(bookRocket(props.id));
  });

  it('should call leaveRocket action when the "Unbook Rocket" button is clicked', () => {
    props.reserved = true;
    component = create(<Rockets {...props} />);
    const button = component.root.findByType('button');
    button.props.onClick();
    expect(dispatch).toHaveBeenCalledWith(leaveRocket(props.id));
  });
});
