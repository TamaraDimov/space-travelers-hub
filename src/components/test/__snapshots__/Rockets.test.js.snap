// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets component should render correctly 1`] = `
<main>
  <div
    className="card"
  >
    <div>
      <img
        alt="rocket"
        className="image"
        src="https://spacex.com/falcon-1.png"
      />
    </div>
    <div
      className="options"
    >
      <h3
        className="name"
      >
        Falcon 1
      </h3>
      <h4 />
      <p
        className="description>"
      >
        The Falcon 1 was a small rocket designed for small payloads and low Earth orbits.
      </p>
      <button
        className="reserved book"
        onClick={[Function]}
        type="button"
      >
        Book Rocket
      </button>
    </div>
  </div>
</main>
`;
